File: backend/bikePartsHub/src/main/java/com/bphTeam/bikePartsHub/service/AppointmentService.java
package com.bphTeam.bikePartsHub.service;

import com.bphTeam.bikePartsHub.dto.pagenated.PaginatedAppointmentResponseDto;
import com.bphTeam.bikePartsHub.dto.request.appointmentRequestDto.AppointmentSaveRequestDto;
import com.bphTeam.bikePartsHub.dto.response.appointmentResponseDto.AppointmentResponseDto;
import com.bphTeam.bikePartsHub.entity.Appointment;
import com.bphTeam.bikePartsHub.entity.ServiceType;
import com.bphTeam.bikePartsHub.utils.AppointmentStatus;

import java.time.LocalDate;
import java.util.List;

public interface AppointmentService {
    List<ServiceType> getAllServices();
    ServiceType getServiceById(Long id);
    List<Appointment> getAppointmentsByDate(LocalDate date);
    void createAppointment(AppointmentSaveRequestDto appointmentDto);
    boolean isTimeSlotAvailable(LocalDate date, String startTime, int duration);
    List<String> getAvailableTimeSlots(LocalDate date, int duration);

    List<AppointmentResponseDto> getCustomerAppointments(Integer id);

    PaginatedAppointmentResponseDto getAllAppointmentDetails(int page, int size);

    String changeAppointmentStatus(long appointmentId, AppointmentStatus status);
}

File: frontend/BikePartsHubApp/src/app/user/pages/appintment-page/appointment.service.ts
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { catchError, from, Observable, switchMap, throwError } from 'rxjs';
import { AuthService } from '../../auth/auth.service';



export interface TimeSlotStatus {
  slot: string;
  status: 'available' | 'busy' | 'highly-busy' | 'not-available';
  bookedCount: number;
}
export interface ServiceType {
  id: number;
  serviceName: string;
  serviceDuration: number;
  serviceCost: number;
  description: string;
  features: string[];
}
export interface Appointment {
  id?: number;
  customerName: string;
  mobile: string;
  plateNumber: string;
  startDate: string;
  startTime: string;
  engineOil: string;
  engineOilCost: number;
  totalCharge: number;
  appointmentStatus: AppointmentStatus;
  serviceType: ServiceType;
  user?: any; // Replace with proper User interface if available
}
export enum AppointmentStatus {
  UPCOMING = 'UPCOMING',
  COMPLETED = 'COMPLETED',
  ATTENDED = 'ATTENDED',
  MISSED = 'MISSED',
}
export interface AppointmentSaveRequest {
  customerName: string;
  mobile: string;
  startDate: string;
  startTime: string;
  plateNumber: string;
  engineOil: string;
  engineOilCost: number;
  totalCharge: number;
  serviceTypeId: number;
  appointmentStatus: AppointmentStatus;
  userId: number;
}
@Injectable({
  providedIn: 'root',
})
export class AppointmentService {
  private readonly API_URL = 'https://bikepartshub.altero.dev/api/v1/appointment';

  constructor(private http: HttpClient, private authService: AuthService) {}

  getAllServices(): Observable<ServiceType[]> {
    return this.http.get<ServiceType[]>(`${this.API_URL}/services`);
  }

  getServiceById(id: number): Observable<ServiceType> {
    return this.http.get<ServiceType>(`${this.API_URL}/services/${id}`);
  }

  getAppointmentsByDate(date: string): Observable<Appointment[]> {
    const params = new HttpParams().set('date', date);
    return this.http.get<Appointment[]>(`${this.API_URL}/appointments`, {
      params,
    });
  }

  getAvailableTimeSlots(date: string, duration: number): Observable<string[]> {
    const params = new HttpParams()
      .set('date', date)
      .set('duration', duration.toString());
    return this.http.get<string[]>(`${this.API_URL}/time-slots`, { params });
  }

  createAppointment(appointment: AppointmentSaveRequest): Observable<void> {
    const createRequest = () => {
      return this.http.post<void>(`${this.API_URL}/appointments`, appointment, {
        headers: this.authService.getAuthHeader(),
      });
    };

    return createRequest().pipe(
      catchError((error) => {
        if (error.status === 401) {
          return this.authService.refreshToken().pipe(
            switchMap(() => createRequest()),
            catchError((refreshError) => {
              return throwError(() => refreshError);
            })
          );
        }
        return throwError(() => error);
      })
    );
  }
}
